symengine/ruby/ext/symengine/symengine_macros.c

```diff
@@ -0,0 +1,51 @@
+#include "symengine_macros.h"
+#include "symengine.h"
+
+void sympify(VALUE operand2, basic_struct **cbasic_operand2) {
+    basic_struct *num_basic, *den_basic;
+    VALUE new_operand2, num, den;
+
+    switch(TYPE(operand2)) {
+        case T_FIXNUM:
+        case T_BIGNUM:
+            new_operand2 = cbasic_alloc(c_integer);
+            Data_Get_Struct(new_operand2, basic_struct, *cbasic_operand2);
+            GET_SYMINTFROMVAL(operand2, *cbasic_operand2);
+            break;
+
+        case T_RATIONAL:
+            num = rb_funcall(operand2, rb_intern("numerator"), 0, NULL);
+            den = rb_funcall(operand2, rb_intern("denominator"), 0, NULL);
+
+            num_basic = basic_new_heap();
+            den_basic = basic_new_heap();
+
+            GET_SYMINTFROMVAL(num, num_basic);
+            GET_SYMINTFROMVAL(den, den_basic);
+
+            new_operand2 = cbasic_alloc(c_rational);
```

Comment:
Ruby is dynamically-typed and the binary functions need to do different things based on the type of the operand they receive. This led to the need for a function like `sympify`. So, that whatever the type of the operand it gets, it extracts the proper pointer from it, either by *unwrapping* or by creating a new Ruby-SymEngine object from the operand and then *unwrapping* it.

So, `sympify` should not need to return the object to ruby. However, we still need an appropriate C struct made from the information that we got from the second operand. And making that ourselves will add an extra overhead to delete that. And it can be tricky many times(like in case of `SymEngine::Integer` in ruby) when we need to make a new object of it and use that for operations. So, there is indeed a need to pass that newly made object to ruby, so that it is properly handled.

What you said about first allocating and then passing it to the function and then again deleting it after getting it back can be done too. Only that we will have to change that for all the functions, that uses sympify.